### ISSUE TITLE
Mautic admin is unusable by blind marketers using a screenreader

### ISSUE THREAD

**Issue by RCheesley** [mautic/mautic#10442]
### Mautic Version

4.0.0

### PHP version

7.4

### What browsers are you seeing the problem on?

Chrome

### What happened?

Posting from a report on the forums:

So I am blind, and use a screen reader to access my computer. The admin pages in mautic seem to use some heavy javascript based combo box replacement, which doesn’t work at all with screen readers. The screen reader sees a item that says it can be clicked with the current setting, then a text box. However, I can’t type in the text box and select anything, and screen reader have a very hard problem clicking things virtually. I posted a bug on github, but I figured I’d post here to see if anyone can think of a workaround or quick patch.

These errors are showing in the log: Not log related.

Steps I have tried to fix the problem:

1. Turn off css to show the original combo box, the combo box shows items, but updating it doesn’t update the javascript box, so mautic thinks nothing has been selected.
2. Find, grep, vim, and echo through lots of files with javascript of css content, trying to find the offending select code and neutralize it, to no avail.

### How can we reproduce this issue?

Access Mautic with a screen reader and try to use the interface

### Code of Conduct

- [X] I confirm that I have read and agree to follow this project's Code of Conduct

**Comment by mautibot** [comment]
This issue has been mentioned on **Mautic Community Forums**. There might be relevant details there:

https://forum.mautic.org/t/for-blind-screen-reader-users-all-combo-boxes-in-the-mautic-admin-are-inaccessible/20892/3

**Comment by RCheesley** [comment]
See also: https://github.com/mautic/mautic/issues/10430

**Comment by alanhartless** [comment]
Hi! I just wanted to quickly add some context for developers for when we look into resolving this. Mautic uses a slightly modified version of [Chosen](https://github.com/harvesthq/chosen) (search for "mautic hack" in the [source](https://github.com/mautic/mautic/blob/0abd6b79738f7cf096b435d8f87ea998617e17c1/app/bundles/CoreBundle/Assets/js/libraries/3a.chosen.jquery.js)) for the combo box user experience for searching/ajaxified dropdown lists. It looks like they have had an open PR for many years attempting to make it accessible. https://github.com/harvesthq/chosen/pull/2047. Since Chosen is no longer maintained, we should consider replacing it with an alternative that has accessibility built in or modify the currently included source to implement accessibility.

**Comment by RCheesley** [comment]
See also a rather lengthy discussion in the Drupal issue queue: https://www.drupal.org/node/1271622 which was morphed into this equally long thread: https://www.drupal.org/project/drupal/issues/2346973#comment-9388809.

Useful comparison spreadsheet: https://docs.google.com/spreadsheets/d/1vifOGL_4wgOl6sy2Os59k24saIHvwFFQoi5brH5Uv28/edit?usp=sharing

**Comment by bbenjamin** [comment]
@RCheesley contacted me to see if I could provide some input on this as a Drupal accessibility maintainer. 

Although I don't think Chosen can be overridden to meet all the structural expectations of a modern accessible UI widget, there are absolutely things you can do to make it usable for screenreader users. The end result may not be ideal, but the experience will resemble "this feature seems a bit outdated" vs the current "this feature doesn't work at all...". I.e. nonsighted users will be able to use it with reasonable ease -- but it won't quite match the experience of something that was built ground-up with accessibility in mind.

A very brief version of my recommendations to make this screenreader-acceptable (and some of this may be conclusions already reached by your team, but just in case...)
1 - On pages with Chosen, if there isn't one already, add a `<div>` with the `aria-live="polite"` attribute/value. Any time the text in this div changes, its contents will be conveyed to screenreaders.
2 - Using something like https://alphagov.github.io/accessible-autocomplete/examples/ as reference, actions in chosen need to be sent to the aria live region from step 1, so when something like an option is selected, the aria region says "option [optionname] selected". Hopefully, Chosen has a rich enough event system that most of this can be done with event listeners. When there isn't an event, consider overriding Chosen functions and have them include calls to update the aria region. If overriding functions doesn't work, Mutation observers can work as well. Any action performed within Chosen that doesn't result in a screenreader doing something probably needs to be addressed.

That is a very truncated version, but hopefully that is enough to let you know it's feasible to make something acceptable with the current library, although far from ideal. If this is the approach you want to go, there are additional considerations worth knowing, but I'll spare you that until it's decided this is worth pursuing.

**Comment by lilmike** [comment]
Any news on this? :-)

**Comment by lilmike** [comment]
As far as I know this is still an issue -- I haven't heard anything to
the contrary.
-Michael.





-- 
Sorry about the test, it's just a me thing.
Michael Taboada, Creator of Games. Does that make me a god, well, sorta.
My pgp key: 79BC390E8864CD71305D4DD606CD2197A6EE5ED7
My website: https://michaels.world
2MB website: https://2mb.games
Don't believe everything you read on the internet. I might not be real, you might not be real, and this email certainly isn't real.
Sent from my smart home...
stationary bicycle

**Comment by lilmike** [comment]
I still haven't heard much on this, just bumping it in case there's any
updates.




-- 
Sorry about the test, it's just a me thing.
Michael Taboada, Creator of Games. Does that make me a god, well, sorta.
My pgp key: 79BC390E8864CD71305D4DD606CD2197A6EE5ED7
My website: https://michaels.world
2MB website: https://2mb.games
Don't believe everything you read on the internet. I might not be real, you might not be real, and this email certainly isn't real.
Sent from my christmas spirited...
macbook pro

**Comment by RCheesley** [comment]
Thanks for bumping it @lilmike - I have been in discussions with the folks at Acquia who are working on accessibility and it is likely that we have some resources coming our way in the coming months to both improve the current situation and help our developers to write more accessible code. I will make sure that this does not drop off the radar.

**Comment by RCheesley** [comment]
Bump

**Comment by RCheesley** [comment]
Not stale

**Comment by lilmike** [comment]
@RCheesley any news on this? Can this be reopened? It's been quite a while... :-(

**Comment by RCheesley** [comment]
We have a small team forming in the UX UI channel on Slack and this will be one of the priorities!

**Comment by RCheesley** [comment]
@andersonjeccel some of the info on this issue may be helpful for your work!

**Comment by lilmike** [comment]
Still relevant.

Btw how's the work progressing?

-Michael.

**Comment by lilmike** [comment]
still valid.

-Michael.

-- 
Sorry about the test, it's just a me thing.
Michael Taboada, Creator of Games. Does that make me a god, well, sorta.
My pgp key: 79BC390E8864CD71305D4DD606CD2197A6EE5ED7
My website: https://michaels.world
2MB website: https://2mb.games
Don't believe everything you read on the internet. I might not be real, you might not be real, and this email certainly isn't real.
Sent from my just now shipped...
stationary bicycle

**Comment by lilmike** [comment]
Any updates on this?

-Michael.





-- 
Sorry about the test, it's just a me thing.
Michael Taboada, Creator of Games. Does that make me a god, well, sorta.
My pgp key: 79BC390E8864CD71305D4DD606CD2197A6EE5ED7
My website: https://michaels.world
2MB website: https://2mb.games
Don't believe everything you read on the internet. I might not be real, you might not be real, and this email certainly isn't real.
Sent from my newfangled...
stationary bicycle
